mender-update/update_module/v3/platform/boost_filesystem/fs_operations.cpp

```diff
@@ -227,6 +250,159 @@ expected::ExpectedStringVector DiscoverUpdateModules(const conf::MenderConfig &c
+	}
+
+	p = fs::path(update_module_workdir_) / "streams";
+	fs::remove_all(p, ec);
+	if (ec) {
+		return error::Error(ec.default_error_condition(), "Could not remove " + p.string());
+	}
+
+	return error::NoError;
+}
+
+AsyncFifoOpener::AsyncFifoOpener(events::EventLoop &loop) :
+	event_loop_(loop),
+	cancelled_(make_shared<atomic<bool>>(true)) {
+}
+
+AsyncFifoOpener::~AsyncFifoOpener() {
+	Cancel();
+}
+
+error::Error AsyncFifoOpener::AsyncOpen(const string &path, ExpectedWriterHandler handler) {
+	// Excerpt from fifo(7) man page:
+	// ------------------------------
+	// The FIFO must be opened on both ends (reading and writing) before data can be passed.
+	// Normally, opening the FIFO blocks until the other end is opened also.
+	//
+	// A process can open a FIFO in nonblocking mode. In this case, opening for read-only
+	// succeeds even if no one has opened on the write side yet and opening for write-only fails
+	// with ENXIO (no such device or address) unless the other end has already been opened.
+	//
+	// Under Linux, opening a FIFO for read and write will succeed both in blocking and
+	// nonblocking mode. POSIX leaves this behavior undefined.  This can be used to open a FIFO
+	// for writing while there are no readers available.
+	// ------------------------------
+	//
+	// We want to open the pipe to check if a process is going to read from it. But we cannot do
+	// this in a blocking fashion, because we are also waiting for the process to terminate,
+	// which happens for Update Modules that want the client to download the files for them. So
+	// we need this AsyncFifoOpener class, which does the work in a thread.
+
+	if (!*cancelled_) {
+		return error::Error(
+			make_error_condition(errc::operation_in_progress), "Already running AsyncFifoOpener");
+	}
+
+	*cancelled_ = false;
+	path_ = path;
+	thread_ = thread([this, handler]() {
+		auto writer = make_shared<events::io::AsyncFileDescriptorWriter>(event_loop_);
+		// This will block for as long as there are no FIFO readers.
+		auto err = writer->Open(path_);
+
+		auto cancelled = cancelled_;
+		if (err != error::NoError) {
+			event_loop_.Post([handler, err, cancelled]() {
+				if (!*cancelled) {
+					// Needed because capture is always const, and `unexpected`
+					// wants to `move`.
+					auto err_copy = err;
+					handler(expected::unexpected(err_copy));
+				}
+			});
+		} else {
+			event_loop_.Post([handler, writer, cancelled]() {
+				if (!*cancelled) {
+					handler(writer);
+				}
+			});
+		}
+	});
+
+	return error::NoError;
+}
+
+void AsyncFifoOpener::Cancel() {
+	if (*cancelled_) {
+		return;
+	}
+
+	*cancelled_ = true;
+
+	// Open the other end of the pipe to jerk the first end loose.
+	int fd = ::open(path_.c_str(), O_RDONLY | O_NONBLOCK);
+	if (fd < 0) {
+		// Should not happen.
+		int errnum = errno;
+		log::Error(string("Cancel::open() returned error: ") + strerror(errnum));
+		assert(fd >= 0);
+	}
+
+	thread_.join();
+
+	::close(fd);
```

Comment:
Q: No need to care about errors here?

Reply:
Generally no. There is not much that can go wrong when closing a file handle, and the best we could do about it is a log entry. We have not typically handled this in Golang either, even though it also provides an error return.